account_fiscal_position_partner_type/models/account_fiscal_position.py

```diff
@@ -0,0 +1,119 @@
+# Copyright 2021 Valentin Vinagre <valentin.vinagre@sygel.es>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import api, fields, models
+
+
+class AccountFiscalPosition(models.Model):
+    _inherit = "account.fiscal.position"
+
+    fiscal_position_type = fields.Selection(
+        selection=[("b2c", "Customer"), ("b2b", "Company")],
+        string="Type",
+        default=lambda self: self._default_fiscal_position_type(),
+    )
+
+    @api.model
+    def _default_fiscal_position_type(self):
+        return self.env.company.default_fiscal_position_type
+
+    @api.model
+    def _get_fpos_by_region_and_position_type(
+        self,
+        country_id=False,
+        state_id=False,
+        zipcode=False,
+        vat_required=False,
+        position_type="b2c",
+    ):
+        if not country_id:
+            return False
+        company_id = self.env.context.get("force_company", self.env.company.id)
+        base_domain = [
+            ("auto_apply", "=", True),
+            ("vat_required", "=", vat_required),
+            ("company_id", "in", [company_id, False]),
+            ("fiscal_position_type", "=", position_type),
+        ]
+        null_state_dom = state_domain = [("state_ids", "=", False)]
+        null_zip_dom = zip_domain = [("zip_from", "=", False), ("zip_to", "=", False)]
+        null_country_dom = [
+            ("country_id", "=", False),
+            ("country_group_id", "=", False),
+        ]
+
+        if zipcode:
+            zip_domain = [("zip_from", "<=", zipcode), ("zip_to", ">=", zipcode)]
+
+        if state_id:
+            state_domain = [("state_ids", "=", state_id)]
+
+        domain_country = base_domain + [("country_id", "=", country_id)]
+        domain_group = base_domain + [("country_group_id.country_ids", "=", country_id)]
+
+        # Build domain to search records with exact matching criteria
+        fpos = self.search(domain_country + state_domain + zip_domain, limit=1)
+        # return records that fit the most the criteria, and fallback on
+        # less specific fiscal positions if any can be found
+        if not fpos and state_id:
```

Comment:
I think this part is not correct.

- if not fpos and state_id: search with state_domain and null_zip_dom
- if not fpos and zipcode: search with null_state_dom and zip_domain
- if not fpos: search same domain_country but null_state_dom and  null_zip_dom

```
        if not fpos and state_id:
            fpos = self.search(domain_country + state_domain + null_zip_dom, limit=1)
        if not fpos and zipcode:
            fpos = self.search(domain_country + null_state_dom + zip_domain, limit=1)
        if not fpos:
            fpos = self.search(domain_country + null_state_dom + null_zip_dom, limit=1)

```

Reply:
See https://github.com/OCA/account-fiscal-rule/pull/206#pullrequestreview-689503368 for the solution for not reimplementing the logic. DRY solution.

Reply:
I have already done the changes, I am checking that everything is correct and I do a push the changes. Good suggestion @pedrobaeza  had not fallen into doing so.

Reply:
okey, that looks better